File: packages/js/core/src/algorithms/parse-query.ts
import { QueryInvocations, QueryDocument, Uri, InvokeResult } from "../types";

import { SelectionSetNode, ValueNode } from "graphql";
import { Tracer } from "@polywrap/tracing-js";
import { Result, ResultErr, ResultOk } from "@polywrap/result";

export const parseQuery = Tracer.traceFunc(
  "core: parseQuery",
  (
    uri: Uri,
    doc: QueryDocument,
    variables?: Record<string, unknown>
  ): Result<QueryInvocations<Uri>, Error> => {
    if (doc.definitions.length === 0) {
      return ResultErr(Error("Empty query document found."));
    }

    const queryInvocations: QueryInvocations<Uri> = {};

    for (const def of doc.definitions) {
      if (def.kind !== "OperationDefinition") {
        const error = Error(
          `Unrecognized root level definition type: ${def.kind}\n` +
            "Please use a 'query' or 'mutation' operations."
        );
        return ResultErr(error);
      }

      // Get the method name
      const selectionSet = def.selectionSet;
      const selections = selectionSet.selections;

      if (selections.length === 0) {
        const error = Error(
          "Empty selection set found. Please include the name of a method you'd like to query."
        );
        return ResultErr(error);
      }

      for (const selection of selections) {
        if (selection.kind !== "Field") {
          const error = Error(
            `Unsupported selection type found: ${selection.kind}\n` +
              "Please query a method."
          );
          return ResultErr(error);
        }

        const method = selection.name.value;
        const invocationName = selection.alias ? selection.alias.value : method;

        if (queryInvocations[invocationName]) {
          const error = Error(
            `Duplicate query name found "${invocationName}". Please use GraphQL aliases that each have unique names.`
          );
          return ResultErr(error);
        }

        // Get all arguments
        const selectionArgs = selection.arguments;
        const args: Record<string, unknown> = {};

        if (selectionArgs) {
          for (const arg of selectionArgs) {
            const name = arg.name.value;

            if (args[name]) {
              return ResultErr(Error(`Duplicate arguments found: ${name}`));
            }

            const extractionResult = extractValue(arg.value, variables);
            if (!extractionResult.ok) {
              return extractionResult;
            }
            args[name] = extractionResult.value;
          }
        }

        queryInvocations[invocationName] = {
          uri,
          method,
          args,
        };
      }
    }

    return ResultOk(queryInvocations);
  }
);

const extractValue = Tracer.traceFunc(
  "core: extractValue",
  (
    node: ValueNode,
    variables?: Record<string, unknown>
  ): InvokeResult<unknown> => {
    if (node.kind === "Variable") {
      // Get the argument's value from the variables object
      if (!variables) {
        const error = Error(
          `Variables were not specified, tried to resolve variable from query. Name: ${node.name.value}\n`
        );
        return ResultErr(error);
      }

      if (variables[node.name.value] === undefined) {
        return ResultErr(Error(`Missing variable: ${node.name.value}`));
      }

      return ResultOk(variables[node.name.value]);
    } else if (
      node.kind === "StringValue" ||
      node.kind === "EnumValue" ||
      node.kind === "BooleanValue"
    ) {
      return ResultOk(node.value);
    } else if (node.kind === "IntValue") {
      return ResultOk(Number.parseInt(node.value));
    } else if (node.kind === "FloatValue") {
      return ResultOk(Number.parseFloat(node.value));
    } else if (node.kind === "NullValue") {
      return ResultOk(null);
    } else if (node.kind === "ListValue") {
      const length = node.values.length;
      const list = [];

      for (let i = 0; i < length; ++i) {
        const extractionResult = extractValue(node.values[i], variables);
        if (!extractionResult.ok) {
          return extractionResult;
        }
        list.push(extractionResult.value);
      }

      return ResultOk(list);
    } else if (node.kind === "ObjectValue") {
      const length = node.fields.length;
      const object: Record<string, unknown> = {};

      for (let i = 0; i < length; ++i) {
        const field = node.fields[i];
        const extractionResult = extractValue(field.value, variables);
        if (!extractionResult.ok) {
          return extractionResult;
        }
        object[field.name.value] = extractionResult.value;
      }

      return ResultOk(object);
    } else {
      return ResultErr(Error(`Unsupported value node: ${node}`));
    }
  }
);

export const extractSelections = Tracer.traceFunc(
  "core: extractSelections",
  (node: SelectionSetNode): Result<Record<string, unknown>, Error> => {
    const result: Record<string, unknown> = {};

    for (const selection of node.selections) {
      if (selection.kind !== "Field") {
        const error = Error(
          `Unsupported result selection type found: ${selection.kind}`
        );
        return ResultErr(error);
      }

      const name = selection.name.value;

      if (result[name]) {
        return ResultErr(Error(`Duplicate result selections found: ${name}`));
      }

      if (selection.selectionSet) {
        const selectionsResult = extractSelections(selection.selectionSet);
        if (!selectionsResult.ok) {
          return selectionsResult;
        }
        result[name] = selectionsResult.value;
      } else {
        result[name] = true;
      }
    }

    return ResultOk(result);
  }
);

File: packages/py/core/src/algorithms/parse_query.py
from __future__ import annotations

from typing import Any, Dict

from graphql import SelectionSetNode, ValueNode

from ..types import InvokeApiOptions


def parse_query(uri: Uri, doc: QueryDocument, variables=None) -> InvokeApiOptions:
    if variables is None:
        variables = {}
    if len(doc.definitions) == 0:
        raise ValueError("Empty query document found.")

    query_invocations = {}
    for definition in doc.definitions:
        if definition.kind != "operation_definition":
            raise ValueError(
                f"""Unrecognized root level definition type: {definition.kind}\n
                Please use a 'query' or 'mutation' operations."""
            )

        selection_set = definition.selection_set
        selections = selection_set.selections

        if len(selections) == 0:
            raise ValueError("Empty selection set found. Please include the name of a method you'd like to query.")

        for selection in selections:
            if selection.kind != "field":
                raise ValueError(f"""Unsupported selection type found: {selection.kind}\nPlease query a method.""")

            method = selection.name.value
            invocation_name = selection.alias.value if selection.alias and selection.alias.value else method

            if query_invocations.get(invocation_name):
                raise ValueError(
                    f"""Duplicate query name found "{invocation_name}".
                    Please use GraphQL aliases that each have unique names."""
                )

            selection_args = selection.arguments
            input = {}

            if selection_args:
                for arg in selection_args:
                    name = arg.name.value

                    if input.get(name):
                        raise ValueError(f"Duplicate input argument found: {name}")

                    input[name] = extract_value(arg.value, variables)

            selection_results = selection.selection_set
            result_filter = None

            if selection_results:
                result_filter = extract_selections(selection_results)

            query_invocations[invocation_name] = InvokeApiOptions(uri, method, input, result_filter)
    return query_invocations


def extract_value(node: ValueNode, variables: Dict[str, Any] = {}) -> Any:
    if node.kind == "variable":
        # Get the argument's value from the variables object
        if not variables:
            raise ValueError(
                f"Variables were not specified, tried to resolve variable from query. Name: {node.name.value}\n"
            )

        if variables.get(node.name.value) is None:
            raise ValueError(f"Missing variable: {node.name.value}")

        return variables[node.name.value]
    elif node.kind == "string_value" or node.kind == "enum_value" or node.kind == "boolean_value":
        return node.value
    elif node.kind == "int_value":
        return int(node.value)
    elif node.kind == "float_value":
        return float(node.value)
    elif node.kind == "null_value":
        return None
    elif node.kind == "list_value":
        length = len(node.values)
        list = []
        for i in range(length):
            list.append(extract_value(node.values[i], variables))
        return list
    elif node.kind == "object_value":
        length = len(node.fields)
        object = {}
        for i in range(length):
            field = node.fields[i]
            object[field.name.value] = extract_value(field.value, variables)
        return object
    else:
        raise ValueError(f"Unsupported value node: {node} and {node.kind}")


def extract_selections(node: SelectionSetNode) -> Dict[str, Any]:
    result = {}
    for selection in node.selections:
        if selection.kind != "field":
            raise ValueError(f"Unsupported result selection type found: {selection.kind}")
        name = selection.name.value

        if result.get(name):
            raise ValueError(f"Duplicate result selections found: {name}")

        if selection.selection_set:
            result[name] = extract_selections(selection.selection_set)
        else:
            result[name] = True

    return result
